src/size.c

```diff
@@ -0,0 +1,346 @@
+#include "vctrs.h"
+#include "utils.h"
+
+// Initialised at load time
+SEXP vec_slice_dispatch_fn = NULL;
+
+// Defined below
+SEXP vec_as_index(SEXP i, SEXP x);
+static void slice_copy_attributes(SEXP to, SEXP from, SEXP index);
+
+
+static void stop_bad_index_length(R_len_t data_n, R_len_t i) {
+  Rf_errorcall(R_NilValue,
+               "Can't index beyond the end of a vector.\n"
+               "The vector has length %d and you've tried to subset element %d.",
+               data_n, i);
+}
+
+#define SLICE(RTYPE, CTYPE, DEREF, CONST_DEREF, NA_VALUE)       \
+  const CTYPE* data = CONST_DEREF(x);                           \
+  R_len_t data_n = Rf_length(x);                                \
+                                                                \
+  R_len_t n = Rf_length(index);                                 \
+  int* index_data = INTEGER(index);                             \
+                                                                \
+  SEXP out = PROTECT(Rf_allocVector(RTYPE, n));                 \
+  CTYPE* out_data = DEREF(out);                                 \
+                                                                \
+  for (R_len_t i = 0; i < n; ++i, ++index_data, ++out_data) {   \
+    int j = *index_data;                                        \
+    *out_data = (j == NA_INTEGER) ? NA_VALUE : data[j - 1];     \
+  }                                                             \
+                                                                \
+  UNPROTECT(1);                                                 \
+  return out
+
+static SEXP lgl_slice(SEXP x, SEXP index) {
+  SLICE(LGLSXP, int, LOGICAL, LOGICAL_RO, NA_LOGICAL);
+}
+static SEXP int_slice(SEXP x, SEXP index) {
+  SLICE(INTSXP, int, INTEGER, INTEGER_RO, NA_INTEGER);
+}
+static SEXP dbl_slice(SEXP x, SEXP index) {
+  SLICE(REALSXP, double, REAL, REAL_RO, NA_REAL);
+}
+static SEXP cpl_slice(SEXP x, SEXP index) {
+  SLICE(CPLXSXP, Rcomplex, COMPLEX, COMPLEX_RO, vctrs_shared_na_cpl);
+}
+static SEXP chr_slice(SEXP x, SEXP index) {
+  SLICE(STRSXP, SEXP, STRING_PTR, STRING_PTR_RO, NA_STRING);
+}
+static SEXP raw_slice(SEXP x, SEXP index) {
+  SLICE(RAWSXP, Rbyte, RAW, RAW_RO, 0);
+}
+
+#undef SLICE
+
+
+#define SLICE_BARRIER(RTYPE, GET, SET, NA_VALUE)                \
```

Comment:
What does barrier stand for?

Reply:
And is it really worth generalising for a single case?

Reply:
It stands for memory barrier, that's for type for which we don't have access to the C array.

This will be useful for altrep vectors so we can index them without materialising them. This would also allow us to reimplement character slicing with the memory barrier accessors in case they deprecate `STRING_PTR`. In the mean time, using `STRING_PTR` is better for performance (and is safe because we only assign CHARSXPs to a fresh vector that belongs to the youngest GC generation).